gensim/topic_coherence/text_analysis.py

```diff
@@ -300,18 +300,10 @@ def accumulate(self, texts, window_size):
     def _iter_texts(self, texts):
         dtype = np.uint16 if np.iinfo(np.uint16).max >= self._vocab_size else np.uint32
         for text in texts:
-            if self.text_is_relevant(text):
-                yield np.fromiter((
-                    self.id2contiguous[self.token2id[w]] if w in self.relevant_words
-                    else self._none_token
-                    for w in text), dtype=dtype, count=len(text))
-
-    def text_is_relevant(self, text):
-        """Check if the text has any relevant words."""
-        for word in text:
-            if word in self.relevant_words:
-                return True
-        return False
+            yield np.fromiter((
+                self.id2contiguous[self.token2id[w]] if w in self.relevant_words
```

Comment:
What happens when there are no relevant words in the text at all? (Why was the `text_is_relevant` check there in the first place?)

Technical note: can you please split up this statement, it is hard to read. E.g. let's give the inner generator its own named variable:

```python
ids = (self.id2contiguous[self.token2id[w]] if w in self.relevant_words else self._none_token for w in text)
yield np.fromiter(ids, …)
```

I realize the ugliness comes from the original code (that you just copied), but it's externally contributed code so let's clean it up while we're at it. Thanks!

Reply:
Actually, looking at that whole `fromiter()` construct – what on earth is that? Why not simply:

```python
ids = [self.id2contiguous[self.token2id[w]] if w in self.relevant_words else self._none_token for w in text]
yield np.array(ids, dtype=dtype)
```

Reply:
@piskvorky np.fromiter() seems to be faster than allocating the entire list and calling np.array(). To investigate, I set up [this Jupyter notebook](https://gist.github.com/aaronjbecker/362c1b88368bccd501a077e0515542f1) to simulate the use case in this text_analysis module and found that using np.fromiter with a generator instead of np.array with a list results in about a ~10-15% speedup. 

The relative performance gain increases with document count and decreases with document length (proportional to share of total work devoted to array creation). I feel like a 10-15% runtime difference is noticeable for coherence evaluation on large corpora?

Readability can still be improved--you can bind the ids variable to a generator expression instead of putting the generator directly in the fromiter call w/o any performance impact. I think a compromise makes sense here--create a separate ids variable for clarity, as a generator, but keep np.fromiter for speed?

@silviatti sorry to hijack your commit thread, wanted to help move this along if possible. I'm curious to see whether this change resolves a difference in Gensim's coherence scores vs. another topic modeling library that I use...

Reply:
Personally I'd go with readability instead of a 15% speed increase. But it's a minor point, both constructs seem fine.

@silviatti can you answer my questions above? We're planning a new release soon, I'd love to get this clean-up merged in. Thanks.

Reply:
@silviatti ping.